### PR TITLE
Improve canonicalPath method in URIUtilTest.java

### PR DESCRIPTION
**Pull request by thesnowgoose**
I noticed the canonicaPath method was so complicated to understand and had duplicated code, so I improved it and now is shorter and runs faster

### CODE REVIEW

**Comment by gregw** [comment]
Leonardo,

simpler code and faster execution are both very good to have - specially in methods like this that are called very frequently.

But how did you verify that it was faster?     Did you test that against the common case of a uri that does not contain any . or .. elements?  

I'm a little dubious because your version is always faster as it will need to compile a regex and creates a lot more objects (arrays etc.).   It may be best to combine the approaches - have a simple contains test for "/." and if that is not present, then return the original URI,  only if we have verified that work needs to be done do we split filter and join?

Also need CLA to be signed for the eclipse process to allow us to merge... plus your { placement does not comply with the formatting standards.

**Comment by cesartg** [comment]
Hi, Leonardo.

Your proposed solution is shorter (and therefore more _digestible_) than the current one.

I have some suggestions for you:

- @gregw has a good point. You should check if the given input contains  '.'  to avoid no needed operations (splitting, iteration and joining). You could add this checking to the first statement from your method:

```
        if (path == null || path.isEmpty() || !path.contains(".")) {
            return path;
        }
```

- An array and two lists are being created. To avoid a list creation, you could do the following (also, you could use a LinkedList instead of an ArrayList. A simple iteration is being performed and the remove operation is faster):

```
        List<String> directories = new LinkedList<>();
        Collections.addAll(directories, path.split("(?<=\\/)"));
```

- The removeIf statement could be removed. (If you remove `dir.isEmpty() ||`, the test still passes), and  you can handle the ¨.¨ case in the following lines.

- A way to simplify the code inside the for loop could be the following:


```
            switch (iterator.next()) {
                case "./":
                case ".":
                    if (iterator.hasNext() && directories.get(iterator.nextIndex()).equals("/")) {
                        break;
                    }
                    iterator.remove();
                    break;
                case "../":
                case "..":
                    if(iterator.previousIndex() == 0) {
                        return null;
                    }
                    iterator.remove();
                    iterator.previous();
                    iterator.remove();
                    break;
            }
```

In that way, we are doing more explicit that if we found a "." we just remove it, and if we find  a "..", we remove the previous one directory, too.  The checking that handle the case when we removed all the directories from the path but the given input path starts with "/" could be outside the loop.

- To join the list elements, you could use `String.join("", directories)` or using a StringBuilder (the internal implementation of String.join uses a StringBuilder).

I applied this changes to your code and I contrasted with the current one. I executed the unit test for the canonical method. In my computer, the current method is a little faster but the time difference was not significant (0.003 seconds approximately). I prefer your version because it is easier to understand.

**Comment by thesnowgoose** [comment]
@gregow you were right, I completely forgot about that scenario where there
is no . or .. elements. Let me check again and I will tell you what I found.
I already signed the CLA and also let me check those standards. Thank both
of you guys.

**Comment by thesnowgoose** [comment]
Hey @gregw, I checked what you said and it is done. Could you check it one more time??
Execution time are almost the same but the code is simpler now (I verified it running the test and comparing execution times). Thanks!!!!

**Comment by cesartg** [comment]
Impressive change, Leonardo.
- It is shorter (30 additions, 116 deletions)
- It is clearer (The cases when you are removing parts of the given path are more explicit)

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 1]
```diff
@@ -503,130 +507,40 @@ public static String parentPath(String p)
...
-                  if (path.charAt(start+1)!='.' || path.charAt(start+2)!='.')
-                      break;
-                  break search;
-            }
-            
-            end=start;
-            start=path.lastIndexOf('/',end-1);
-        }
+        List<String> directories = new LinkedList<>();
+        Collections.addAll(directories, path.split("(?<=\\/)"));
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: One final change.... could you make this a static precompiled Pattern  eg
```` java
   static Pattern __PATH_SPLIT = Pattern.compile("(?<=\\/)");
````
and  then this line becomes the more efficient
````java
    Collections.addAll(directories, __PATH_SPLIT.split(path);
````

I'm pretty sure that Pattern is threadsafe in this way.

**Comment by gregw** [comment]
Looking pretty good.  Just see my line note about precompiling the regex and I think we are good.

Note that you don't need to close/open the PR.  Just push your commits to your branch and we will note them here.

**Comment by thesnowgoose** [comment]
It is done. Thank you

**Comment by gregw** [comment]
Agh!  The merge broke the build!

Test org.eclipse.jetty.server.HttpConnectionTest.testBadPathDotDotPath() is failing because the path "/ooops/../../path"   returns "path" when it should return null.

Can you add a path like this to the URIUtilTest and we can then work out what is going on....

If this can't be fixed quickly, then we will have to revert.

**Comment by gregw** [comment]
I have reverted and added a unit test

**Comment by thesnowgoose** [comment]
I already fix it, I am testing it right now

**Pull request merged**